h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -2725,10 +2912,10 @@ private void doCompute() {
       }
       if (_state._iter >= _parms._max_iterations)
         _job.warn("Reached maximum number of iterations " + _parms._max_iterations + "!");
-      if (_parms._nfolds > 1 && !Double.isNaN(_lambdaCVEstimate))
-        _model._output.setSubmodelIdx(_model._output._best_submodel_idx = _bestCVSubmodel);  // reset best_submodel_idx to what xval has found
+      if (_parms._nfolds > 1 && !Double.isNaN(_lambdaCVEstimate) && _bestCVSubmodel < _model._output._submodels.length)
```

Comment:
Unfortunately, this is needed. I found out about this when running `pyunit_PUBDEV_7366_gam_cross_validation_fold_assignment_modulo_multinomial_large` - there was a difference when running with just one alpha or multiple alphas. I thought there is some information leak between each alpha runs but I found out there is not (at least not a relevant one). This test didn't use to fail because the winning alpha wasn't wining (`for (; lidx < lmin_max; ++lidx) ` where lmin_max was the max id of winning model which ended up being excluded) and the other issue was that `bestCVSubmodel` didn't really point to the submodel that we would like to as it can take more (and maybe sometimes less) iterations for the main model to create a submodel. So I am wondering if using lambda search with submodel index is valid as it can point to a different lambda value than the winning one.

Reply:
@tomasfryda In my view, what you have done is the following:

1.  Go through model from each fold to find the average test deviance;
2. If uses set lambda and alpha, model will just run main model on the best test deviance with whatever alpha and lambda values it found during the cv_compute thingy.
3. if user uses lambda_search, you will again find the best average test deviance and find what alpha and lambda it founds it at.  Then the main model will run just that one alpha and a sequence of lambda starting from the max lambda until you hit the test lambda and may be go one lambda lower.

So you are saying doing it this way does not work?

Reply:
In all these cases I am assuming we are doing CV. (Runs without CV shouldn't change)
(1) and (2) work exactly as you described. 

(3) worked and works differently.

**How it used to work:**

cv_compute picks the best alpha and lambda parameters from submodels 0...best_submodel_index (best submodel is excluded) and gets the index of CV[0].submodel that should have those alpha and lambda value. Then in doCompute we run GLM with the same max_iteration constraint and hope to train the right model which doesn't have to happen (especially if the picked submodel is the last one) - the main model has more data to train on and it can take more iterations to converge and that can influence the number of submodels that are trained.
Then we pick some model with index `_bestCVSubmodel ` that can have different alpha and lambda than the one that was best even if we use `lambda_search`.

**How it should work now:**

If the `lambda_search==false`, we pick the best alpha and lambda and train just one submodel for the main model. 

If we have `lambda_search==true`, it is pretty similar as it used to behave but it eliminates the possibility of using different than selected alpha value as we train only the submodels for the correct alpha.

Then we pick some model with index `_bestCVSubmodel ` if it exists, if it doesn't we `pickBestModel`. If the submodel exists we should be guaranteed that it has the same alpha and lambda value as the best submodel from the cv_compute thingy.